### ISSUE TITLE
[Whitelist] Build Service plugin support request.

### ISSUE THREAD

**Issue by ChadoriRebornXD** [Scirra/Construct-3-bugs#3130]
Hi. I would like to request support for a few plugins, to wit:

**Update**
- cordova-plugin-inappbrowser (3.1.0)
- cordova-plugin-facebook4 (5.0.0) :
   Param 1 : "APP_ID"
   Param 2 : "APP_NAME"
- cordova-plugin-googleplus (7.0.2) :
   Param 1 : "REVERSED_CLIENT_ID"
   Param 2 : "WEB_APPLICATION_CLIENT_ID"
- cordova-plugin-x-toast (2.7.2)
- cordova-plugin-dialogs (2.0.2)
- cordova-plugin-badge (0.8.8)

**Add**
- cordova-plugin-device (2.0.3)
- cordova-android-support-gradle-release (3.0.1)
- cordova-plugin-enable-multidex (0.2.0)
- cordova-plugin-admob-free (0.27.0)
   Param 1 : "ADMOB_APP_ID"
- cordova-admob-sdk (0.24.1)
   Param 1 : "PLAY_SERVICES_VERSION"
- cordova-plugin-background-mode-2 (0.8.1)

**Add#2**
- lycwed-cordova-plugin-admob-applovin (1.0.1)
- lycwed-cordova-plugin-admob-chartboost (1.0.1)
- lycwed-cordova-plugin-admob-facebook (1.0.3)
- cordova-plugin-admob-mediation (1.1.2)
- lycwed-cordova-plugin-admob-inmobi (1.0.9)
- cordova-admob-mobfox (0.0.3)
- lycwed-cordova-plugin-admob-unityads (1.0.2)
- lycwed-cordova-plugin-admob-vungle (1.0.21)
  Param 1 : "REWARDVIDEO_ID"
  Param 2 : "INTERSTITIAL_ID"

**Add#3**
- cordova-plugin-websocket-server (1.5.0)
- cordova-plugin-zeroconf (1.4.0)

**For consideration**
_Just in case it's already feasible to support.._
- cordova-plugin-local-notifications@0.9.0-beta.2
- pushwoosh-cordova-plugin@7.17.0

---------------------------------------------------------

I finally want all my addons to get supported by the **C3 Build Service** and update the existing ones, so that the users, including me, won't have to always rely on PhoneGap and be able to build directly.

I hope the long list is alright with you @shortercode 😁 .

**Thank you.**

**Comment by shortercode** [comment]
`cordova-plugin-googleplus@5.3.1` is already on the whitelist, do you specifically need the older version as well? The latest version is 7.0.2

**Comment by ChadoriRebornXD** [comment]
Ohh, this is a typo. Sorry.
I meant, `cordova-plugin-googleplus@7.0.2`.
Edited.

**Comment by ChadoriRebornXD** [comment]
Hi. Is this good to go? 🙂

**Comment by shortercode** [comment]
I'm about halfway through the list, had a few things on last week and had
Friday off. Will let you know next week hopefully

**Comment by ChadoriRebornXD** [comment]
@shortercode it's fine. Thank you very much!

**Comment by KekGames** [comment]
@shortercode do you plan to update and add everything at once, or maybe plugins that you have already reviewed could be updated earlier?

**Comment by shortercode** [comment]
I've updated the whitelist for the following plugins so far:
- cordova-plugin-inappbrowser@3.1.0
- corodva-plugin-facebook4@5.0.0
- cordova-plugin-x-toast@2.7.2
- cordova-plugin-dialogs@2.0.2
- cordova-plugin-badge@0.8.7
- cordova-plugin-device@2.0.3
- cordova-plugin-background-mode-2@0.8.1

**Comment by dop2000** [comment]
@shortercode , Thanks for doing this! Could you please give an estimate on when these two plugins will be added?
cordova-plugin-admob-free (0.27.0)
cordova-admob-sdk (0.24.1)

**Comment by shortercode** [comment]
I've added a further 3 to the whitelist:
- cordova-admob-mobfox@0.0.3
- cordova-plugin-websocket-server@1.5.0
- cordova-plugin-zeroconf@1.4.0

@dop2000 it's probably going to be a few weeks, due to staff leave. There's a large number of the admob related plugins which are potentially dangerous, so may not pass. We're investigating if we can make adjustments to safely use them.

**Comment by ChadoriRebornXD** [comment]
@shortercode Thank you very much, hoping for the others too.

**Comment by ChadoriRebornXD** [comment]
Hi @shortercode , I plan to write a couple of mediation Cordova plugins.

What determines compatibility with the **C3 Build Service**? I remember you mentioning restricting some dependencies for some unpredictable behavior.

Do I need to look out for something? Thanks.

**Comment by JeFawk** [comment]
Having these supported would be a great addition!

**Comment by shortercode** [comment]
It's not so much about compatibility, they will work but we cannot consider them safe. The review basically covers 2 main restrictions:

1. Dependencies are reproducible - i.e. no dependency will automatically update, with the exception of trusted sources ( i.e. play-games-services )
2. Only simple build scripts with obvious meaning are accepted

It's normally the first which is the one that plugins get held up on. It's not a particularly restrictive policy, but perhaps not one which 3rd party developers aim for.

**Comment by ChadoriRebornXD** [comment]
@shortercode Ohh I see. I understand.

On our part, Cordova plugins are time consuming to develop and maintain, hence there are barely any publicly available ones for each feature, and the authors are not aware of the **Construct Build Service**'s requirements. 
We are only lucky If someone already has made one for each feature, but if it turns out to have non-straightforward scripts or dependencies, then that feature would be out of our reach. And I don't think they would bother to change their plugin on our behalf.

We really would try to adjust and create our own Cordova plugins on some implementations practical for us, but it's nearly impossible to replicate some features like services or redo already well-established plugins.

I hope there was an easier way for this, where we can take advantage of the little available plugins we could use, but I understand that this for the safety and reliability of the build service.

Thanks for the information.

-----

@shortercode If it's alright, I would like to request the change of version of the plugin:
- `plugin name="cordova-admob-sdk" spec="~0.13.1"`
  - Param 1 : `PLAY_SERVICES_VERSION`

This is for the compatibility of Construct 3's **Google Play** plugin, thank you again!

**Comment by shortercode** [comment]
I agree that it's problematic finding suitable cordova plugins for features. The plugin ecosystem is a bit of a mess, even with "good" plugins maintenance is slow and many are either abandoned, outdated or just badly written. We're now maintaining several plugins and have plans for more, but they are quite time consuming to keep up to date.

Yesterday I came up with a potential way of allowing us to tweak each plugin slightly, so that they meet the requirements. I'm trying to decide the feasibility of it at the moment. But it wouldn't guarantee compat. with every plugin.

**Comment by ChadoriRebornXD** [comment]
> We're now maintaining several plugins and have plans for more, but they are quite time consuming to keep up to date.

> Yesterday I came up with a potential way of allowing us to tweak each plugin slightly, so that they meet the requirements. I'm trying to decide the feasibility of it at the moment. But it wouldn't guarantee compat. with every plugin.

That's great thanks.

**Comment by ChadoriRebornXD** [comment]
Hi, would like to request an update for a plugin:
- cordova-plugin-admob-mediation (2.0.0)

Thanks.

**Comment by shortercode** [comment]
Will add it onto the list.

New system is coming along well, hoping to have it on production this week.

**Comment by ChadoriRebornXD** [comment]
@shortercode That's great news! Thanks!

**Comment by ChadoriRebornXD** [comment]
Hi, also `pushwoosh-cordova-plugin@7.18.1` updated to a new version for an important fix. Thanks again!

**Comment by ChadoriRebornXD** [comment]
I nearly forgot : `pushwoosh-geozones-cordova-plugin@1.17.0`. Thanks again!

**Comment by shortercode** [comment]
The dependency upgrade is complete, however we have been delayed by a Cordova bug that has reared it's head with the new system. We had previously worked around it but apparently it has become a problem again. As the upgrade was quite a time sink I have put it aside for the time being. But I'm hoping to return to it once some other high priority issues have been resolved.

On a positive note the new system has allowed several plugins which were deemed risky under the old system to pass security review. I'm hoping it will also allow for faster builds, but further testing will be required for that.

**Comment by ChadoriRebornXD** [comment]
@shortercode It's alright and thank you for working on it!

Also, I would like to request an update to the Facebook addon.
`cordova-plugin-facebook4 : @6.0.0`
- Param 1 : "APP_ID"
- Param 2 : "APP_NAME"
- Param 3 : "FACEBOOK_HYBRID_APP_EVENTS"

Thanks!

**Comment by ChadoriRebornXD** [comment]
Also these:
`cordova-plugin-badge@0.8.8`
`cordova-plugin-android-notch@1.0.3`
`cordova-plugin-cutout@1.0.10`

Sorry for the workload, but we're really enjoying Construct. 🙂
Thanks again.

**Comment by masscube** [comment]
Can you add this too, please?
https://www.npmjs.com/package/cordova-plugin-email
Thx :)

**Comment by ERAgames777** [comment]
@shortercode Hi, I bought plugins from chadory mobile-ads but they do not work on c3build output, the creator says to contact you, can you do this somehow so that c3build can recognize them during construction?

**Comment by ChadoriRebornXD** [comment]
Hi, I need to update this plugin's version `pushwoosh-cordova-plugin@7.18.3`. Thanks.

**Comment by ERAgames777** [comment]
When will admob mediation plugins and intermediary advertising networks plugins work with c3build?

**Comment by ChadoriRebornXD** [comment]
If it's already feasible:
`onesignal-cordova-plugin`@latest_version_when_implemented

Thanks.

**Comment by masscube** [comment]
@shortercode Is there a list what plugins are now supported?

**Comment by shortercode** [comment]
@masscube we don't have a public list at this time, I'll bring it up with the team so we can discuss it.

**Comment by ChadoriRebornXD** [comment]
`cordova-plugin-network-information@latest_version_when_implemented`, thanks!

**Comment by shortercode** [comment]
A public list is now available on the SDK forum https://www.construct.net/en/forum/construct-3/plugin-sdk-10/build-server-cordova-plugin-149266

**Comment by masscube** [comment]
@shortercode can you add **cordova-plugin-email @latest_version** please!

**Comment by shortercode** [comment]
While I appreciate that this bug is listed in a very general way, and has yet to be closed, I would prefer if requests could go in new issues. This way we can track which ones have been completed.

**Comment by masscube** [comment]
@shortercode Ok, here my new request: https://github.com/Scirra/Construct-3-bugs/issues/3583

**Comment by shortercode** [comment]
cordova-plugin-network-information@2.0.2 has been added to the whitelist.

**Comment by shortercode** [comment]
onesignal-cordova-plugin@2.8.1 has been added to the whitelist

**Comment by shortercode** [comment]
I've been trying to go back through this list thread, but it's grown unwieldy. As per my comment to @masscube it's better to post requests as new issues rather than comments so that it's easier for us to manage.

@ChadoriRebornXD could you confirm which versions of plugins you are still waiting on?

**Comment by shortercode** [comment]
@ERAgames777 I believe the mediation and lycwed plugins failed the security review.

Internally we have an enhancement project that we hoped would enable us to be less strict on security reviews, but it has hit a number of deadends due to unfortunate behaviours in the cordova toolchain. We haven't cancelled it, but it's on hold at the moment. That project was the primary reason this issue was kept open, instead of just listing a dozen or so plugins as having failed review.

I understand that you would like us to support them but crossposting about it in other issues will not enable us to support them any faster.

**Comment by shortercode** [comment]
@ChadoriRebornXD I reviewed pushwoosh-cordova-plugin@7.18.5 ( latest) and it passed. Are you happy with that version, or do you specifically want 7.18.3?

**Comment by ChadoriRebornXD** [comment]
@shortercode 7.18.5 is perfect, thanks!

**Comment by shortercode** [comment]
pushwoosh-cordova-plugin@7.18.5 has been added to the whitelist

**Comment by shortercode** [comment]
I'm we're going to close this issue. We've added all the ones that passed review to the whitelist. If anyone would like to request a particular plugin is added to the whitelist then please file a new report. I've marked this report with a new label to indicate that their are some plugins listed here that couldn't be added. In future if plugins cannot be added to the list for some reason it should be marked with this label.

If we make any changes in the future that reduce the restrictions then we can revisit the issues closed with this label.

**Comment by ChadoriRebornXD** [comment]
@shortercode Not much we can do if it can't be passed. I'll adjust for now and find some possible alternatives. Thanks for the hard work on the currently passed ones and the list.
